### PR TITLE
core/thread: thread block legality check

### PR DESCRIPTION
**Pull request by ZetaR60**
### Contribution description
This PR adds a simple sanity check to mbox, msg, mutex, etc. that triggers a kernel panic when a core function tries to block the current thread while within an ISR. The intention is to produce a clear failure compared to the current behavior which can trigger problems that are harder to diagnose. I believe that this will be helpful because there are a lot of libraries that call these core functions and it may not be immediately obvious to less experienced users which ones are ISR safe and which ones are not.

### Issues/PRs references
#10488

### CODE REVIEW

**Comment by jcarrano** [comment]
I'm not sure I understand the issue, but I see a pattern of 
    `sched_set_status(blocked or sleeping) -> irq_restore(...) -> thread_yield()`

Could this be turned into a function of its own, that also includes this check?

**Comment by kaspar030** [comment]
This needs to be guarded (e.g., by DEVELHELP). I'd rather not bloat deployed code with runtime checks, in the hottest code paths.

Maybe simple asserts are sufficient?

**Comment by ZetaR60** [comment]
> Maybe simple asserts are sufficient?

From what I understand, assert is usually used to verify that parameters fulfill their documented restrictions, so it might be clearer to have a separate error message. But if you think it is better to use assert, I can change it.

> I see a pattern of
> ```sched_set_status(blocked or sleeping) -> irq_restore(...) -> thread_yield()```
> Could this be turned into a function of its own, that also includes this check?

Not everything fits this pattern exactly, and you would be passing a lot of parameters to the function.

**Comment by kaspar030** [comment]
> From what I understand, assert is usually used to verify that parameters fulfill their documented restrictions, so it might be clearer to have a separate error message. But if you think it is better to use assert, I can change it.

TL;DR Yes, please.

The function panics if called within an ISR, with a proper error message. It is very useful when developing, but not worth the overhead (runtime and flash) in production code. assert does exactly that.

The only difference of the added function to a plain assert is a distinct panic code and and a distinct error message. assert also points to the exact position in the binary.

I don't see any use for the panic code. 
A message can be appended like this: ```assert(!irq_is_in() && "function must not be called from within ISR!");```.

The naming of the function is misleading. "thread_block_legal_check()" panics when called from ISR. I wouldn't assume that when just looking at the function name. "panic_if_called_from_isr()" would be much more to the point. Then "assert(!irq_is_in())" is much more idiomatic and used like that in many other places in RIOT's codebase.

**Comment by ZetaR60** [comment]
@kaspar030 I have changed things to follow your suggestions.
